shell.nix

```diff
@@ -21,13 +22,39 @@ shared.pkgs.runCommand "dhall-shell-${shell}" {
     echo "    Overriding option completion shell flavor:"
     echo "        $ nix-shell --argstr shell zsh"
 
-    source <(dhall            --${shell}-completion-script dhall)
-    source <(dhall-to-bash    --${shell}-completion-script dhall-to-bash)
-    source <(dhall-to-nix     --${shell}-completion-script dhall-to-nix)
-    source <(dhall-to-json    --${shell}-completion-script dhall-to-json)
-    source <(dhall-to-yaml    --${shell}-completion-script dhall-to-yaml)
-    source <(dhall-to-yaml-ng --${shell}-completion-script dhall-to-yaml-ng)
-    source <(json-to-dhall    --${shell}-completion-script json-to-dhall)
-    source <(yaml-to-dhall    --${shell}-completion-script yaml-to-dhall)
+    EXECUTABLES=(
+      dhall
+      dhall-to-bash
+      dhall-to-nix
+      dhall-to-json
+      dhall-to-yaml
+      dhall-to-yaml-ng
+      json-to-dhall
+      yaml-to-dhall
+    )
+
+    if [ "${shell}" == "zsh" ]; then
+      # zsh does not support loading completions using `source`, so we need to
+      # add a special case for zsh
+
+      COMPLETIONS=$(mktemp -d)
+      export ZDOTDIR=$(mktemp -d)
+      trap "rm --recursive -- $COMPLETIONS $ZDOTDIR" INT EXIT
```

Comment:
I'm just curious: I suspect this tries to run the `rm` command, but "catches the exception" when it fails?

Reply:
@sjakobi: Could you clarify what you mean?

Reply:
I've just never seen `trap` before. My guess was that it's a bit like

    <command> || true

…but then you'd probably have used that idiom.

I've just run `man trap` on my system, which starts with

```
trap(1)                                                                                            fish                                                                                            trap(1)

NAME
       traptrap - perform an action when the shell receives a signal
```

I assume on `zsh` it does something similar.

Reply:
@sjakobi not necessarily when it fails - `INT` is in case it's interrupted and `EXIT` is for when the process exits normally. Think of it more like `defer` in Go (if you're familiar with Go).

Reply:
Thanks @ari-becker! I was reminded of `defer`, but I had forgotten what it's called. It's been a while, since I last looked a Go!

Reply:
Yeah, you can think of `trap` as a poor man's `finally` for the shell (since it doesn't really have a properly safe `bracket`)